// dll-webpack-plugin/lib/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
exports.PLUGIN_NAME = "dllWebpackPlugin";
exports.DLL_VERSION_JSON_NAME = "dll.version.json";
exports.DEFAULT_PACKAGE_JSON_PATH = "./package.json";
exports.NODE_ENV = process.env.NODE_ENV || "production";
exports.PLUGIN_PATH = path.resolve("./" + exports.PLUGIN_NAME);
exports.MANIFEST_JSON_PATH = path.resolve("./" + exports.PLUGIN_NAME + "/" + exports.NODE_ENV + "/[name].manifest.json");
exports.DLL_VERSION_JSON_PATH = path.resolve("./" + exports.PLUGIN_NAME + "/" + exports.DLL_VERSION_JSON_NAME);
exports.DLL_PATH = path.resolve("./" + exports.PLUGIN_NAME + "/" + exports.NODE_ENV);
exports.isProduction = exports.NODE_ENV === "production";
exports.DEFAULT_LIBRARY = "[name]_library_dll";
exports.defaultDllVersionJson = {
    "name": exports.PLUGIN_NAME,
    "onceForceUpdate": false,
    "alwayForceUpdate": false,
    "dllConfig": {
        "development": 0,
        "production": 0
    },
    "dllEntry": 0
};
